iroh-net/src/relay/client.rs

```diff
@@ -253,17 +253,87 @@ impl<W: AsyncWrite + Unpin + Send + 'static> ClientWriter<W> {
 /// The Builder returns a [`Client`] starts a [`ClientWriter`] run task.
 pub struct ClientBuilder {
     secret_key: SecretKey,
-    reader: WebSocketRead<MaybeTlsStreamReader>,
-    writer: WebSocketWrite<MaybeTlsStreamWriter>,
-    local_addr: SocketAddr,
+    reader: RelayConnReader,
+    writer: RelayConnWriter,
+    local_addr: Option<SocketAddr>,
+}
+
+#[derive(derive_more::Debug)]
+pub(crate) enum RelayConnReader {
+    Relay(
+        #[debug("FramedRead<MaybeTlsStreamReader, DerpCodec>")]
+        FramedRead<MaybeTlsStreamReader, DerpCodec>,
+    ),
+    Ws(#[debug("SplitStream<WebSocketStream>")] SplitStream<WebSocketStream>),
+}
+
+#[derive(derive_more::Debug)]
+pub(crate) enum RelayConnWriter {
+    Relay(
+        #[debug("FramedWrite<MaybeTlsStreamWriter, DerpCodec>")]
+        FramedWrite<MaybeTlsStreamWriter, DerpCodec>,
+    ),
+    Ws(
+        #[debug("SplitSink<WebSocketStream, tokio_tungstenite_wasm::Message>")]
+        SplitSink<WebSocketStream, tokio_tungstenite_wasm::Message>,
+    ),
+}
+
+fn tung_to_io_err(e: tokio_tungstenite_wasm::Error) -> std::io::Error {
+    std::io::Error::new(std::io::ErrorKind::Other, e.to_string())
+}
+
+impl Stream for RelayConnReader {
+    type Item = Result<Frame>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        match *self {
+            Self::Relay(ref mut ws) => ws.poll_next_unpin(cx),
```

Comment:
please use `Pin::new(ws).poll_next(cx)` for these, to keep in line with the rest of the code